iib/workers/tasks/opm_operations.py

```diff
@@ -190,3 +191,201 @@ def _serve_cmd_at_port(serve_cmd, cwd, port, max_tries, wait_time):
+
+    if dockerfile_name:
+        df_path_tmp = os.path.join(base_dir, dockerfile_name)
+        os.rename(df_path, df_path_tmp)
+        df_path = df_path_tmp
+
+    if index_db:
+        db_path = get_worker_config()['hidden_index_db_path']
+        rel_path_index_db = os.path.relpath(index_db, base_dir)
+        with open(df_path, 'a') as f:
+            f.write(f'\nADD {rel_path_index_db} {db_path}\n')
+
+    log.info("Dockerfile was successfully generated.")
+    return df_path
+
+
+@retry(exceptions=IIBError, tries=2, logger=log)
+def _opm_registry_add(
+    base_dir,
+    index_db,
+    bundles,
+    from_index=None,
+    overwrite_from_index_token=None,
+    overwrite_csv=False,
+    container_tool=None,
+):
+    """
+    Add the input bundles to an operator index database.
+
+    This only runs operations on index database.
+
+    :param str base_dir: the base directory to generate the database and index.Dockerfile in.
+    :param str index_db: relative path to SQLite index.db database file
+    :param list bundles: a list of strings representing the pull specifications of the bundles to
+        add to the index image being built.
+    :param str from_index: the pull specification of the container image containing the index that
+        the index image build will be based from.
+    :param str overwrite_from_index_token: the token used for overwriting the input
+        ``from_index`` image. This is required to use ``overwrite_from_index``.
+        The format of the token must be in the format "user:password".
+    :param bool overwrite_csv: a boolean determining if a bundle will be replaced if the CSV
+        already exists.
+    :param str container_tool: the container tool to be used to operate on the index image
+    """
+    from iib.workers.tasks.utils import set_registry_token, run_cmd
+
+    # The bundles are not resolved since these are stable tags, and references
+    # to a bundle image using a digest fails when using the opm command.
+    bundle_str = ','.join(bundles) or '""'
+
+    cmd = [
+        'opm',
+        'registry',
+        'add',
+        '--database',
+        index_db,
+        # This enables substitutes-for functionality for rebuilds. See
+        # https://github.com/operator-framework/enhancements/blob/master/enhancements/substitutes-for.md
+        '--enable-alpha',
+        '--bundle-images',
+        bundle_str,
+    ]
+
+    if container_tool:
+        cmd.append('--container-tool')
+        cmd.append(container_tool)
+
+    log.info('Generating the database file with the following bundle(s): %s', ', '.join(bundles))
+
+    if overwrite_csv:
+        log.info('Using force to add bundle(s) to index')
+        cmd.extend(['--overwrite-latest'])
+
+    with set_registry_token(overwrite_from_index_token, from_index):
+        run_cmd(cmd, {'cwd': base_dir}, exc_msg='Failed to add the bundles to the index image')
+
+
+@retry(exceptions=IIBError, tries=2, logger=log)
+def opm_registry_add_fbc(
```

Comment:
In the near future, if you think this file is becoming too large, consider moving fbc functions for opm to a separate module. Please not my approval is not contingent on this change. I am simply suggesting something that can be kept in mind :)

Reply:
Yes I was considering to move this to `fbc_utils` ... but I am not sure if I would not complicate imports once again.